readme.txt

additonal stream quality choice: HD allows 720p playback
TVGuide addon support is through .strm files
Format of a .strm file is:
plugin://plugin.video.ustvnow/?name=ABC&mode=play

Where name is the channel name given by USTVNow in exactly the same case as given in the menu.
&mode=play tells the addon to play the channel

see the .strm files in the plugin.video.ustvnowtvguide directory for examples

changelog.txt
version 0.3.1 - includes .strm support

default.py
'''
    ustvnow XBMC Plugin
    Copyright (C) 2011 t0mm0

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

from resources.lib import Addon, ustvnow 
import sys
import urllib
import xbmc, xbmcgui, xbmcplugin

Addon.plugin_url = sys.argv[0]
Addon.plugin_handle = int(sys.argv[1])
Addon.plugin_queries = Addon.parse_query(sys.argv[2][1:])

email = Addon.get_setting('email')
password = Addon.get_setting('password')
ustv = ustvnow.Ustvnow(email, password)

Addon.log('plugin url: ' + Addon.plugin_url)
Addon.log('plugin queries: ' + str(Addon.plugin_queries))
Addon.log('plugin handle: ' + str(Addon.plugin_handle))

mode = Addon.plugin_queries['mode']

if mode == 'main':
    Addon.log(mode)
    Addon.add_directory({'mode': 'live'}, Addon.get_string(30001))
    Addon.add_directory({'mode': 'recordings'}, Addon.get_string(30002))

elif mode == 'live':
    Addon.log(mode)
    stream_type = ['rtmp', 'rtsp'][int(Addon.get_setting('stream_type'))]
    channels = ustv.get_channels(int(Addon.get_setting('quality')), 
                                     stream_type)
    for c in channels:
        Addon.add_video_item(c['url'],
                             {'title': '%s - %s' % (c['name'], 
                                                    c['now']['title']),
                              'plot': c['now']['plot']},
                             img=c['icon'])

elif mode == 'recordings':
    Addon.log(mode)
    stream_type = ['rtmp', 'rtsp'][int(Addon.get_setting('stream_type'))]
    recordings = ustv.get_recordings(int(Addon.get_setting('quality')), 
                                     stream_type)
    for r in recordings:
        cm_del = (Addon.get_string(30003), 
                  'XBMC.RunPlugin(%s/?mode=delete&del=%s)' % 
                       (Addon.plugin_url, urllib.quote(r['del_url'])))
        title = '%s (%s on %s)' % (r['title'], r['rec_date'], r['channel'])
        Addon.add_video_item(r['stream_url'], {'title': title, 
                                               'plot': r['plot']},
                             img=r['icon'], cm=[cm_del], cm_replace=True)

elif mode == 'delete':
    dialog = xbmcgui.Dialog()
    ret = dialog.yesno(Addon.get_string(30000), Addon.get_string(30004), 
                       Addon.get_string(30005))
    if ret == 1:
        ustv.delete_recording(Addon.plugin_queries['del'])

elif mode=='play':
    name = Addon.plugin_queries['name']
    print "Play USTVNOW"
    try:
       stream_type = 'rtmp'
       channels = []
       quality = int(Addon.get_setting('quality'))
       channels = ustv.get_channels(quality,stream_type)
       for c in channels:
         if c['name'] == name:
           print "URL = "+c['url']
           print "setResolvedUrl"
           item = xbmcgui.ListItem(path=c['url'])
           xbmcplugin.setResolvedUrl(int(sys.argv[1]), True, item)
    except:
      pass


    
Addon.end_of_directory()
        



resources/lib/ustvnow.py
'''
    ustvnow XBMC Plugin
    Copyright (C) 2011 t0mm0

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''
import Addon
import cookielib
import os
import re
import urllib, urllib2

class Ustvnow:
    __BASE_URL = 'http://lv2.ustvnow.com'
    def __init__(self, user, password):
        self.user = user
        self.password = password
                    
    def get_channels(self, quality=1, stream_type='rtmp'):
        self._login()
        html = self._get_html('iphone_ajax', {'tab': 'iphone_playingnow', 
                                              'token': self.token})
        channels = []
        for channel in re.finditer('class="panel".+?title="(.+?)".+?src="' + 
                                   '(.+?)".+?class="nowplaying_item">(.+?)' +
                                   '<\/td>.+?class="nowplaying_itemdesc".+?' +
                                   '<\/a>(.+?)<\/td>.+?href="(.+?)"',
                                   html, re.DOTALL):
            name, icon, title, plot, url = channel.groups()
            if not url.startswith('http'):
                now = {'title': title, 'plot': plot.strip()}
                if name == 'CW' or name == 'PBS' or name == 'My9':
                  my_quality = quality
                else: my_quality = quality+1
                url = '%s%s%d' % (stream_type, url[4:-1], my_quality)
                channels.append({'name': name, 'url': url, 
                               'icon': icon, 'now': now})
                print "name=<"+name+">/n"
        return channels        

    def get_recordings(self, quality=1, stream_type='rtmp'):
        self._login()
        html = self._get_html('iphone_ajax', {'tab': 'iphone_viewdvrlist'})
        recordings = []
        for r in re.finditer('class="panel".+?title="(.+?)".+?src="(.+?)".+?' +
                             'class="nowplaying_item">(.+?)<\/td>.+?(?:<\/a>' +
                             '(.+?)<\/td>.+?)?vertical-align:bottom.+?">(.+?)' +
                             '<\/div>.+?_self" href="(rtsp.+?)".+?href="(.+?)"', 
                             html, re.DOTALL):
            chan, icon, title, plot, rec_date, url, del_url = r.groups()
            url = '%s%s%s' % (stream_type, url[4:-7], 
                              ['350', '650', '950'][quality])
            if plot:
                plot = plot.strip()
            else:
                plot = ''
            recordings.append({'channel': chan,
                               'stream_url': url,
                               'title': title,
                               'plot': plot,
                               'rec_date': rec_date.strip(),
                               'icon': icon,
                               'del_url': del_url
                               })
        return recordings
    
    def delete_recording(self, del_url):
        html = self._get_html(del_url)
        print html
    
    def _build_url(self, path, queries={}):
        if queries:
            query = Addon.build_query(queries)
            return '%s/%s?%s' % (self.__BASE_URL, path, query) 
        else:
            return '%s/%s' % (self.__BASE_URL, path)

    def _fetch(self, url, form_data=False):
        if form_data:
            Addon.log('posting: %s %s' % (url, str(form_data)))
            req = urllib2.Request(url, form_data)
        else:
            Addon.log('getting: ' + url)
            req = url
        try:
            response = urllib2.urlopen(url)
            return response
        except urllib2.URLError, e:
            Addon.log(str(e), True)
            return False
        
    def _get_html(self, path, queries={}):
        html = False
        url = self._build_url(path, queries)

        response = self._fetch(url)
        if response:
            html = response.read()
        else:
            html = False
        
        return html

    def _login(self):
        Addon.log('logging in') 
        self.token = None
        self.cj = cookielib.CookieJar()
        opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(self.cj))
        
        urllib2.install_opener(opener)
        url = self._build_url('iphone_login', {'username': self.user, 
                                               'password': self.password})
        response = self._fetch(url)
        #response = opener.open(url)
        
        for cookie in self.cj:
            print '%s: %s' % (cookie.name, cookie.value)
            if cookie.name == 'token':
                self.token = cookie.value


resources/lib/Addon.py
'''
    ustvnow XBMC Plugin
    Copyright (C) 2011 t0mm0

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

import cgi
import os.path
import pickle
import re
import sys
import time
import urllib

try:
    import xbmc, xbmcaddon, xbmcgui, xbmcplugin
    is_xbmc = True
except:
    is_xbmc = False
    print 'not running on xbmc'

def log(msg, err=False):
    return
    if err:
        xbmc.log(addon.getAddonInfo('name') + ': ' + msg.encode('utf-8'), 
                 xbmc.LOGERROR)    
    else:
        xbmc.output(addon.getAddonInfo('name') + ': ' + msg.encode('utf-8'), 
                    xbmc.LOGDEBUG)    

def show_error(details):
    show_dialog(details, get_string(30000), True)

def show_dialog(details, title='ustvnowtvguide', is_error=False):
    error = ['', '', '']
    text = ''
    for k, v in enumerate(details):
        error[k] = v
        text += v + ' '
    log(text, is_error)
    dialog = xbmcgui.Dialog()
    ok = dialog.ok(title, error[0], error[1], error[2])
    
def get_setting(setting):
    return addon.getSetting(setting)
    
def get_string(string_id):
    return addon.getLocalizedString(string_id)   

def add_music_item(item_id, infolabels, img='', fanart='', total_items=0):
    infolabels = decode_dict(infolabels)
    url = build_plugin_url({'play': item_id})
    log('adding item: %s - %s' % (infolabels['title'], url))
    listitem = xbmcgui.ListItem(infolabels['title'], iconImage=img, 
                                thumbnailImage=img)
    listitem.setInfo('music', infolabels)
    listitem.setProperty('IsPlayable', 'true')
    listitem.setProperty('fanart_image',fanart)
    xbmcplugin.addDirectoryItem(plugin_handle, url, listitem, 
                                isFolder=False, totalItems=total_items)

def add_video_item(url, infolabels, img='', fanart='', total_items=0, 
                   cm=[], cm_replace=False):
    infolabels = decode_dict(infolabels)
    if url.find('://') == -1:
        url = build_plugin_url({'play': url})
    log('adding item: %s - %s' % (infolabels['title'].decode('utf-8','ignore'), url))
    listitem = xbmcgui.ListItem(infolabels['title'], iconImage=img, 
                                thumbnailImage=img)
    listitem.setInfo('video', infolabels)
    listitem.setProperty('IsPlayable', 'true')
    listitem.setProperty('fanart_image', fanart)
    if cm:
        listitem.addContextMenuItems(cm, cm_replace)
    xbmcplugin.addDirectoryItem(plugin_handle, url, listitem, 
                                isFolder=False, totalItems=total_items)

def add_directory(url_queries, title, img='', fanart='', total_items=0):
    url = build_plugin_url(url_queries)
    log('adding dir: %s - %s' % (title, url))
    listitem = xbmcgui.ListItem(decode(title), iconImage=img, thumbnailImage=img)
    if not fanart:
        fanart = addon.getAddonInfo('path') + '/fanart.jpg'
    listitem.setProperty('fanart_image', fanart)
    xbmcplugin.addDirectoryItem(plugin_handle, url, listitem, 
                                isFolder=True, totalItems=total_items)

def add_artist(artist, total_items=0):
    url_queries = {'mode': 'get_music_directory', 'id': artist['id']}
    add_directory(url_queries, artist['name'], total_items=total_items) 

def add_song(song, img='', total_items=0):
    infolabels = {'title': unicode(song.get('title', get_string(30003))),
                  'artist': unicode(song.get('artist', get_string(30004))),
                  'album': unicode(song.get('album', get_string(30005))),
                  'tracknumber': int(song.get('track', 0)),
                  'genre': unicode(song.get('genre', '')),
                  'duration': int(song.get('duration', 0)),
                 }
    year = song.get('year', None)
    if year:
        infolabels['year'] = year
    add_music_item(song['id'], infolabels, img, total_items)

def add_album(album, img='', total_items=0):
    infolabels = {'title': unicode(album.get('title', get_string(30003))),
                  'artist': unicode(album.get('artist', get_string(30004))),
                 }
    add_directory({'mode': 'get_music_directory', 'id': album['id']}, 
                  album['title'], img, total_items)

def resolve_url(stream_url):
    xbmcplugin.setResolvedUrl(plugin_handle, True, 
                              xbmcgui.ListItem(path=stream_url))

def end_of_directory():
    xbmcplugin.endOfDirectory(plugin_handle)

def build_query(queries):
    return '&'.join([k+'='+urllib.quote(str(v)) for (k,v) in queries.items()])
                                
def build_plugin_url(queries):
    url = plugin_url + '?' + build_query(queries)
    return url

def parse_query(query, clean=True):
    queries = cgi.parse_qs(query)
    q = {}
    for key, value in queries.items():
        q[key] = value[0]
    if clean:
        q['mode'] = q.get('mode', 'main')
        q['play'] = q.get('play', '')
    return q

def show_settings():
    addon.openSettings()

def get_input(title='', default=''):
    kb = xbmc.Keyboard(default, title, False)
    kb.doModal()
    if (kb.isConfirmed()):
        return kb.getText()
    else:
        return False

#http://stackoverflow.com/questions/1208916/decoding-html-entities-with-python/1208931#1208931
def _callback(matches):
    id = matches.group(1)
    try:
        return unichr(int(id))
    except:
        return id

def decode(data):
    return re.sub("&#(\d+)(;|(?=\s))", _callback, data).strip()

def decode_dict(data):
    for k, v in data.items():
        if type(v) is str or type(v) is unicode:
            data[k] = decode(v)
    return data

if is_xbmc:
    addon = xbmcaddon.Addon(id='plugin.video.ustvnowtvguide')
    plugin_path = addon.getAddonInfo('path')

